### DISCUSSION TITLE
`Error: Invariant: cookies() expects to have requestAsyncStorage, none available.` came out when try to use `cookies()` in dynamic routes

### DISCUSSION THREAD

**Discussion by steadily-worked** [vercel/next.js#58639] [Help] [answered]
### Summary

Hi, when i tried to get API call in Server Component using `cookies()` in dynamic routes (like `app/projects/[projectId]/page.tsx`) with `prefetchQuery` of react-quey, `⨯ Error: Invariant: cookies() expects to have requestAsyncStorage, none available.` came out. 
this errors occur only in dynamic routes because when i tried to get same API call in static routes (like `accounts/settings`), this error didn't come out.
i really want to know why this error occurs only in dynamic routes.

### Example

1. page.tsx
```tsx
const Page = async ({
  params,
}: {
  params: { id: string }
}) => {
  const queryClient = new QueryClient()

  await queryClient.prefetchQuery({
    queryKey: [SOME_QUERY_KEY_1],
    queryFn: () => getInformation(params.id),
  })

  return (
    <HydrationBoundary state={dehydrate(queryClient)}>
      <Detail id={params.id} />
    </HydrationBoundary>
  )
}

export default Page
```

2. `getInformation`

```tsx
export const getInformation = async (
  id: string 
): Promise<Information> => {
  try {
    const response = await GET(`${API_ROUTE}/${id}`)
    return response
  } catch (error) {
    throw error
  }
}
```

3. `GET`

```tsx
"use server"

import { decamelizeKeys } from "humps"
import { cookies } from "next/headers"
import { TOKEN_KEY } from "utils/token"

const cookieStore = cookies()
const token = cookieStore.get(TOKEN_KEY)

const headers = {
  "Content-Type": "application/json",
  ...(token && { Authorization: `Bearer ${token.value}` }),
}

export const GET = async (url: string, params?: Record<string, string>) => {
  const queryString = params
    ? new URLSearchParams(decamelizeKeys(params)).toString()
    : ""

  const response = await fetch(queryString ? `${url}?${queryString}` : url, {
    method: "GET",
    cache: "no-cache",
    headers,
  })

  if (!response.ok) {
    console.log("Response Body:", await response.text())
    throw new Error("failed to fetch data")
  }
  return response.json()
}
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Could you try with:

```ts
"use server"

import { decamelizeKeys } from "humps"
import { cookies } from "next/headers"
import { TOKEN_KEY } from "utils/token"

export const GET = async (url: string, params?: Record<string, string>) => {
  const cookieStore = cookies()
  const token = cookieStore.get(TOKEN_KEY)

  const headers = {
    "Content-Type": "application/json",
    ...(token && { Authorization: `Bearer ${token.value}` }),
  }

  const queryString = params
    ? new URLSearchParams(decamelizeKeys(params)).toString()
    : ""

  const response = await fetch(queryString ? `${url}?${queryString}` : url, {
    method: "GET",
    cache: "no-cache",
    headers,
  })

  if (!response.ok) {
    console.log("Response Body:", await response.text())
    throw new Error("failed to fetch data")
  }
  return response.json()
}
```

**Reply by steadily-worked** [reply]
i already did it, and it works! thanks for your help. i think the reason is that in [this page](https://nextjs.org/docs/app/building-your-application/rendering/server-components#server-rendering-strategies), 
> With Dynamic Rendering, routes are rendered for each user at <b>request time</b>.

anyway, thanks :)

**Comment by 6feferonka9** [comment]
Hello, I have the same issue except my flow is different and I feel stuck.

I have a server action, which uses axios instance. This instance reads cookies and sends them to API. Everything works fine in static routes but dynamic are broken.

**Reply by 6feferonka9** [reply]
I solved this by putting cookies().get inside axios req interceptor and not on the top of the file, where I was constructing header before

**Reply by MalvinJay** [reply]
This worked for me!!

**Comment by jmakinin** [comment]
I faced a similar issue recently, All my teammates were using Mac for development, and I only joined later on using a Windows PC. In my case I only had to bump the Next version from "13.4.19"  to "13.5", and for some reason, it just worked fine.
